File: src/main/java/com/example/smarthouse/sensor/light/LightSensor.java
package com.example.smarthouse.sensor.light;

import com.example.smarthouse.sensor.DefaultSensor;
import com.example.smarthouse.sensor.config.SensorProtocol;
import com.example.smarthouse.sensor.config.SensorState;
import com.example.smarthouse.sensor.config.SensorType;
import com.example.smarthouse.sensor.message.MessageType;
import com.example.smarthouse.sensor.message.SensorMessage;
import lombok.Data;

import java.util.Date;

public class LightSensor extends DefaultSensor {
    private boolean lightOn;

    public LightSensor(int sensorId, String sensorName, SensorProtocol sensorProtocol) {
        super(sensorId, sensorName, SensorState.OFF, sensorProtocol, SensorType.SMART_LIGHT);
    }

    public void switchLight() {
        lightOn = !lightOn;
        if (lightOn) {
            setSensorState(SensorState.ON);
            System.out.println("Light turned ON");
        } else {
            setSensorState(SensorState.OFF);
            System.out.println("Light turned OFF");
        }
    }

    @Override
    public SensorMessage getDataFromSensor() {
        return new SensorMessage(
                getSensorId(),
                lightOn ? 1 : 0,
                new Date(),
                MessageType.SENSOR_DATA,
                lightOn ? "Light is ON" : "Light is OFF",
                getSensorType()
        );
    }
}

File: src/main/kotlin/com/example/sensor/LightSensor.kt
package com.example.smarthouse.sensor.light

import com.example.config.SensorProtocol
import com.example.config.SensorState
import com.example.config.SensorType
import com.example.message.MessageType
import com.example.message.Message
import com.example.smarthouse.sensor.DefaultSensor
import java.util.*

class LightSensor(sensorId: Int, sensorName: String?, sensorProtocol: SensorProtocol?) :
    DefaultSensor(sensorId, sensorName, SensorState.OFF, sensorProtocol, SensorType.SMART_LIGHT) {
    var lightOn = false
    fun switchLight() {
        lightOn = !lightOn
        if (lightOn) {
            setSensorState(SensorState.ON)
            println("Light turned ON")
        } else {
            setSensorState(SensorState.OFF)
            println("Light turned OFF")
        }
    }

    override val dataFromSensor: Message
        get() = Message(
            getSensorId(),
            if (lightOn) 1 else 0,
            Date(),
            MessageType.SENSOR_DATA,
            if (lightOn) "Light is ON" else "Light is OFF",
            getSensorType()
        )
}
